helix-core/src/shellwords.rs

```diff
@@ -1,198 +1,476 @@
-    /// #[cfg(unix)]
+    /// assert_eq!(Shellwords::from(":open a\\ ").ends_with_whitespace(), true);
     /// assert_eq!(Shellwords::from(":open a\\ b.txt").ends_with_whitespace(), false);
     /// ```
+    #[inline]
     pub fn ends_with_whitespace(&self) -> bool {
-        matches!(self.state, State::OnWhitespace)
+        self.input.ends_with(' ')
+    }
+}
+
+/// An iterator over an input string which yields arguments.
+///
+/// Splits on whitespace, but respects quoted substrings (using double quotes, single quotes, or backticks).
+/// It also handles lists enclosed in square brackets, treating them as a single argument.
+#[derive(Debug, Clone, Copy)]
+pub struct Args<'a> {
+    input: &'a str,
+    bytes: &'a [u8],
+    idx: usize,
+    start: usize,
+    in_quotes: bool,
+    quote: u8,
+    in_list: bool,
+    is_finished: bool,
+}
+
+impl<'a> Args<'a> {
+    #[inline]
+    fn parse(input: &'a str) -> Self {
+        Self {
+            input,
+            bytes: input.as_bytes(),
+            idx: 0,
+            start: 0,
+            in_quotes: false,
+            quote: b'\0',
+            in_list: false,
+            is_finished: false,
+        }
+    }
+
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.input.is_empty()
+    }
+
+    /// Returns the args exactly as input.
+    ///
+    /// # Examples
+    /// ```
+    /// # use helix_core::shellwords::Args;
+    /// let args = Args::from(r#"sed -n "s/test t/not /p""#);
+    /// assert_eq!(r#"sed -n "s/test t/not /p""#, args.raw());
+    ///
+    /// let args = Args::from(r#"cat "file name with space.txt""#);
+    /// assert_eq!(r#"cat "file name with space.txt""#, args.raw());
+    /// ```
+    #[inline]
+    pub fn raw(&self) -> &str {
+        self.input
+    }
+
+    #[inline(always)]
+    pub const fn empty() -> Self {
+        Self {
+            input: "",
+            bytes: &[],
+            idx: 0,
+            start: 0,
+            in_quotes: false,
+            quote: b'\0',
+            in_list: false,
+            is_finished: true,
+        }
+    }
+}
+
+#[allow(clippy::copy_iterator)]
+impl<'a> Iterator for Args<'a> {
+    type Item = &'a str;
+
+    #[inline]
+    #[allow(clippy::too_many_lines)]
+    fn next(&mut self) -> Option<Self::Item> {
+        /// Walks backwords checking for `\` and counting them to determine if escaped.
+        #[inline]
+        const fn is_escaped(bytes: &[u8]) -> bool {
+            let mut backslashes = 0;
+            let mut idx = bytes.len() - 1;
+
+            loop {
+                if bytes[idx] != b'\\' {
+                    break;
+                }
+                backslashes += 1;
+
+                idx = match idx.checked_sub(1) {
+                    Some(idx) => idx,
+                    None => break,
+                };
+            }
+
+            (backslashes % 2) != 0
+        }
+
+        if self.is_finished {
```

Comment:
This was for simpler use to initiate a termination more explicitly in areas where there could be special casing, such as if a quote is never closed. I can remove this and just refactor to advance the idx so that the while loop terminates instead. Can you give a clear directive to follow for how you want this handled?

Reply:
Did end up removing this, and while I was at it moved some fields to local variables instead, taking the `Args` struct size from 56 bytes to 32.